pallets/stateful-storage/src/types.rs

```diff
@@ -0,0 +1,299 @@
+//! Types for the Stateful Storage Pallet
+use crate::Config;
+use codec::{Decode, Encode, MaxEncodedLen};
+use common_primitives::{
+	msa::MessageSourceId,
+	schema::SchemaId,
+	stateful_storage::{PageHash, PageId},
+};
+use frame_support::pallet_prelude::*;
+use scale_info::TypeInfo;
+use sp_core::bounded::BoundedVec;
+use sp_std::{
+	cmp::*,
+	collections::btree_map::BTreeMap,
+	fmt::Debug,
+	hash::{Hash, Hasher},
+	prelude::*,
+};
+use twox_hash::XxHash64;
+
+/// pallet storage prefix
+pub const PALLET_STORAGE_PREFIX: &[u8] = b"stateful-storage";
+/// itemized storage prefix
+pub const ITEMIZED_STORAGE_PREFIX: &[u8] = b"itemized";
+/// paginated storage prefix
+pub const PAGINATED_STORAGE_PREFIX: &[u8] = b"paginated";
+
+/// MultipartKey type for Itemized storage
+pub type ItemizedKey = (SchemaId,);
+/// MultipartKey type for Paginated storage (full key)
+pub type PaginatedKey = (SchemaId, PageId);
+/// MultipartKey type for Paginated storage (prefix lookup)
+pub type PaginatedPrefixKey = (SchemaId,);
+/// Itemized page type
+pub type ItemizedPage<T> = Page<<T as Config>::MaxItemizedPageSizeBytes>;
+/// Paginated Page type
+pub type PaginatedPage<T> = Page<<T as Config>::MaxPaginatedPageSizeBytes>;
+
+/// Defines the actions that can be applied to an Itemized storage
+#[derive(Clone, Encode, Decode, Debug, TypeInfo, MaxEncodedLen, PartialEq)]
+#[scale_info(skip_type_params(DataSize))]
+#[codec(mel_bound(DataSize: MaxEncodedLen))]
+pub enum ItemAction<DataSize: Get<u32> + Clone + sp_std::fmt::Debug + PartialEq> {
+	/// Adding new Item into page
+	Add {
+		/// The data to add
+		data: BoundedVec<u8, DataSize>,
+	},
+	/// Removing an existing item by index number. Index number starts from 0
+	Delete {
+		/// Index (0+) to delete
+		index: u16,
+	},
+}
+
+/// This header is used to specify the byte size of an item stored inside the buffer
+/// All items will require this header to be inserted before the item data
+#[derive(Encode, Decode, PartialEq, MaxEncodedLen, Debug)]
+pub struct ItemHeader {
+	/// The length of this item, not including the size of this header.
+	pub payload_len: u16,
+}
+
+/// Errors dedicated to parsing or modifying pages
+#[derive(Debug, PartialEq)]
+pub enum PageError {
+	/// Unable to decode the data in the item
+	ErrorParsing(&'static str),
+	/// Add or Delete Operation was not possible
+	InvalidAction(&'static str),
+	/// ItemPage count overflow catch
+	ArithmeticOverflow,
```

Comment:
sp_runtime has ArithmeticError we use it in the MSA pallet.

https://github.com/LibertyDSNP/frequency/blob/db81e4fa142087d8b5cd37b181852f3bb8407803/pallets/msa/src/lib.rs#L985

Reply:
It does but we can't use it here, We want to return All errors under the same Enum

Reply:
I am not sure why we are deviating from what we do everywhere else?

Reply:
Because we are trying to create self contained structs that can be tested in isolation. That's why the parsing logic is implemented here which is inside the types.

Reply:
I think thats good thinking. Something I just noticed is that maybe these errors are not reaching their full potential since they all get swallowed into corrupted state CorruptedState. Maybe we can put this error messages to work. Even if we had an arithemtic error we would not know or a parsing error or anything bad that happens here.

Perhaps it would help us is we had something along the lines of 
```
 pub enum Error<T> {
      CorruptedState(PageError | WhataverError)
     }
```

Reply:
We are printing all the parsing related errors. I think we can print the actions. also but that's less crucial since it will indicate an erroneous input. The parsing problem is way more important that's why we are printing.
```
			ItemizedOperations::<T>::apply_item_actions(&existing_page, &actions[..]).map_err(
				|e| match e {
					PageError::ErrorParsing(err) => {
						log::warn!(
							"failed parsing Itemized msa={:?} schema_id={:?} {:?}",
							state_owner_msa_id,
							schema_id,
							err
						);
						Error::<T>::CorruptedState
					},
					_ => Error::<T>::InvalidItemAction,
				},
			)?;
```